forge-game/src/main/java/forge/game/trigger/Trigger.java

```diff
@@ -184,17 +177,68 @@ public final String replaceAbilityText(final String desc, SpellAbility sa) {
                 sa = wa.getWrappedAbility();
 
                 // wrapped Charm spells are special,
-                // only get the selected abilities
+                // only get the selected abilities if there are any yet
                 if (ApiType.Charm.equals(sa.getApi())) {
                     saDesc = sa.getStackDescription();
+                } else if (ApiType.ImmediateTrigger.equals(sa.getApi())) {
+                    saDesc = sa.toString();
+                    if (sa.hasAdditionalAbility("Execute")) {
+                        SpellAbility overridingSA = sa.getAdditionalAbility("Execute");
+                        if (ApiType.Charm.equals(overridingSA.getApi())) {
+                            saDesc = CharmEffect.makeFormatedDescription(overridingSA, !forStack);
+                        }
+                    }
                 } else {
                     saDesc = sa.toString();
+                    // if ABILITY is used, there is probably Charm somewhere - but at this point no abilities have
+                    // been selected
+                    SpellAbility subSa = sa.getSubAbility();
+                    while (subSa != null) {
+                        if (ApiType.Charm.equals(subSa.getApi())) {
+                            saDesc = CharmEffect.makeFormatedDescription(subSa, !forStack);
+                            break;
+                        } else if (ApiType.ImmediateTrigger.equals(subSa.getApi())) {
+                            if (subSa.hasAdditionalAbility("Execute")) {
+                                SpellAbility overridingSA = subSa.getAdditionalAbility("Execute");
+                                if (ApiType.Charm.equals(overridingSA.getApi())) {
+                                    saDesc = CharmEffect.makeFormatedDescription(overridingSA, !forStack);
+                                    break;
+                                }
+                            }
+                        }
+                        subSa = subSa.getSubAbility();
+                    }
                 }
             } else if (ApiType.Charm.equals(sa.getApi())) {
                 // use special formating, can be used in Card Description
                 saDesc = CharmEffect.makeFormatedDescription(sa);
+            } else if (ApiType.ImmediateTrigger.equals(sa.getApi())) {
```

Comment:
I think with some refactoring and maybe a single flag you would only need both of these new blocks once